ext/ffi/fast_call.rs

```diff
@@ -189,3 +97,373 @@ fn is_fast_api_rv(rv: NativeType) -> bool {
+      }
+      (6, Signed(B)) => {
+        dynasm!(self.assembler; movsx r9d, BYTE [rsp + rsp_offset])
+      }
+      (6, Unsigned(W)) => {
+        dynasm!(self.assembler; movzx r9d, WORD [rsp + rsp_offset])
+      }
+      (6, Signed(W)) => {
+        dynasm!(self.assembler; movsx r9d, WORD [rsp + rsp_offset])
+      }
+      (6, Unsigned(DW) | Signed(DW)) => {
+        dynasm!(self.assembler; mov r9d, [rsp + rsp_offset])
+      }
+      (6, Unsigned(QW) | Signed(QW)) => {
+        dynasm!(self.assembler; mov r9, [rsp + rsp_offset])
+      }
+
+      (_, Unsigned(B)) => dynasm!(self.assembler
+        ; movzx eax, BYTE [rsp + rsp_offset]
+        ; mov [rsp + new_rsp_offset], eax
+      ),
+      (_, Signed(B)) => dynasm!(self.assembler
+        ; movsx eax, BYTE [rsp + rsp_offset]
+        ; mov [rsp + new_rsp_offset], eax
+      ),
+      (_, Unsigned(W)) => dynasm!(self.assembler
+        ; movzx eax, WORD [rsp + rsp_offset]
+        ; mov [rsp + new_rsp_offset], eax
+      ),
+      (_, Signed(W)) => dynasm!(self.assembler
+        ; movsx eax, WORD [rsp + rsp_offset]
+        ; mov [rsp + new_rsp_offset], eax
+      ),
+      (_, Unsigned(DW) | Signed(DW)) => dynasm!(self.assembler
+        ; mov eax, [rsp + rsp_offset]
+        ; mov [rsp + new_rsp_offset], eax
+      ),
+      (_, Unsigned(QW) | Signed(QW)) => dynasm!(self.assembler
+        ; mov rax, [rsp + rsp_offset]
+        ; mov [rsp + new_rsp_offset], rax
+      ),
+    }
+  }
+
+  fn zero_first_arg(&mut self) {
+    dynasm!(self.assembler
+      ; xor rdi, rdi
+    );
+  }
+
+  fn cast_return_value(&mut self, rv: NativeType) {
+    // 8 and 16 bit integers are extended to 32 bits
+    match rv {
+      NativeType::U8 => dynasm!(self.assembler; movzx eax, al),
+      NativeType::I8 => dynasm!(self.assembler; movsx eax, al),
+      NativeType::U16 => dynasm!(self.assembler; movzx eax, ax),
+      NativeType::I16 => dynasm!(self.assembler; movsx eax, ax),
+      _ => (),
+    }
+  }
+
+  fn allocate_stack(&mut self, params: &[NativeType]) {
+    let mut stack_size = {
+      let mut int = -Self::INTEGER_REG;
+      let mut sse = -Self::SSE_REG;
+      for param in params {
+        match param {
+          NativeType::F32 | NativeType::F64 => sse += 1,
+          _ => int += 1,
+        }
+      }
+      ((int.max(0) + sse.max(0)) * 8) as u16
+    };
+
+    // Align stack frame
+    if (stack_size + 8) % 16 != 0 {
+      // Section 3.2.2 of the SysV AMD64 ABI:
+      // > The end of the input argument area shall be aligned on a 16 (32 or 64, if
+      // > __m256 or __m512 is passed on stack) byte boundary. In other words, the value
+      // > (%rsp + 8) is always a multiple of 16 (32 or 64) when control is transferred to
+      // > the function entry point. The stack pointer, %rsp, always points to the end of the
+      // > latest allocated stack frame.
+      stack_size += 8;
+    }
+
+    dynasm!(self.assembler
+      ; sub rsp, stack_size as i32
```

Comment:
Godbolt rustc compiler seems to use push and pop for stack allocation, presumably as long as there is no need to push parameters onto the stack. So this would be the case when the return value has to be cast from 8 or 16 bits to 32 bits.

Would it be worth to add that as a special case, or rather leave such stuff out as micro-optimisation?

The rustc compiler also does some funky things with using movups to (if I understand this correctly) move multiple sequential integer arguments temporarily into xmmN (where N is the number of float args) to make stack arg shifting quicker. But that's the kind of optimisation that is probably entirely unnecessary for us :D

Reply:
> The rustc compiler also does some funky things with using movups to (if I understand this correctly) move multiple sequential integer arguments temporarily into xmmN (where N is the number of float args) to make stack arg shifting quicker. But that's the kind of optimisation that is probably entirely unnecessary for us :D

Invalid unless we assume SSE extension on the target platform.

Reply:
> Godbolt rustc compiler seems to use push and pop for stack allocation, presumably as long as there is no need to push parameters onto the stack. So this would be the case when the return value has to be cast from 8 or 16 bits to 32 bits.

AFAIK Rustc (and many others, with the notable exception of GCC) optimize by using push instead of the normative stack allocation in the prologue as long as no SSE instruction is/can be used. This is because stack instructions (PUSH, POP, CALL, RET) are optimized since the addition of the stack engine in Pentium M and newer CPUs, taking fewer uops than `sub + mov`. Once SSE is used to move packed data, `sub` is needed anyway so I guess the optimization is no longer significant. See https://godbolt.org/z/PeKaxzjK6

> The rustc compiler also does some funky things with using movups to (if I understand this correctly) move multiple sequential integer arguments [...] to make stack arg shifting quicker.

Correct. In a nutshell, it's moving 128 bits of data with a single uops. In the Godbolt linked above, trampoline2 moves parameter k and l together with `movups`.  

>  into xmmN (where N is the number of float args)
no, N is just part of the register name. When passing floats as parameters, N does have some semantic in the sense that the first float goes to XMM0, the second to XMM1, etc. But when moving data in/out of the register as a temporary operation the N does not have any significance. Unless I misunderstood what you meant?

> But that's the kind of optimisation that is probably entirely unnecessary for us

Not sure if necessary or not in the long run, but I do think that this first iteration should focus on a simple and solid implementation, and worry about optimization in the future. It is not like TinyCC applies any of these optimizations anyway. 

> Invalid unless we assume SSE extension on the target platform. (Which is what Bun does, but Deno so far has not, as far as I'm aware)

Keep in mind that this implementation is specific for AMD64, which includes SSE & SSE2 instructions.

Reply:
> AFAIK Rustc (and many others, with the notable exception of GCC) optimize by using push instead of the normative stack allocation in the prologue as long as no SSE instruction is/can be used. This is because stack instructions (PUSH, POP, CALL, RET) are optimized since the addition of the stack engine in Pentium M and newer CPUs, taking fewer uops than `sub + mov`. Once SSE is used to move packed data, `sub` is needed anyway so I guess the optimization is no longer significant. See https://godbolt.org/z/PeKaxzjK6

Alright. Sounds good to me to leave a relatively insignificant complication out.

> no, N is just part of the register name. When passing floats as parameters, N does have some semantic in the sense that the first float goes to XMM0, the second to XMM1, etc. But when moving data in/out of the register as a temporary operation the N does not have any significance. In the previously linked GodBolt, it uses xxm2 because it is the first available xmm register (xmm0 and xmm1) are used for parameters. Unless I misunderstood what you meant?

Yeah I just meant that it's indeed using the first available xmm register: If there are 0 float args then xmm0 can be used, otherwise 1 or 2 or ... based on the number of float params.
 

> Not sure if necessary or not in the long run, but I do think that this first iteration should focus on a simple and solid implementation, and worry about optimization in the future. It is not like TinyCC applies any of these optimizations anyway.

Sounds good to me. Also, I think it's very unlikely we'll run into many C APIs with this many parameters.